### PR TITLE
wayfire: init at 0.6.0

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

I’ve been working this for the past few days independently of #86569, which I just saw. We should compare the two and see if we can combine approaches.

The major difference is that this PR supports Wayfire plugins through a standard Nixpkgs-style `wayfire.withPackages` interface. Upstream Wayfire only works if plugins are installed into Wayfire’s prefix, which isn’t really an option for us, at least with arbitrary plugins. But the good news is they understand the problem, and are willing to accept my patches. So everything that’s a patch in this PR should hopefully be part of the next Wayfire release, at which point these patches can all be dropped.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
I’ve realised that wlc will need to be wrapped too, so I’d like to come up with some nice function that maybe gives you an attrset with wlc and wayfire packages, both wrapped with the right environment variables set.  The hard part is coming up with a name…

**Comment by alyssais** [comment]
> I’ve realised that wlc will need to be wrapped too, so I’d like to
> come up with some nice function that maybe gives you an attrset with
> wlc and wayfire packages, both wrapped with the right environment
> variables set.  The hard part is coming up with a name…

I've now done this:

Wayfire and WCM can both be built with a custom plugin set like so:

    with wayfireApplications.withPlugins (ps: with ps; [ wf-shell ]);
      [ wayfire wcm ]

The default wayfireApplications in Nixpkgs has just wf-shell available.
An unwrapped wayfireApplications is available as
wayfireApplications-unwrapped.

**Comment by wucke13** [comment]
You may add me to the maintainers, if  you'd like to.

**Comment by alyssais** [comment]
> You may add me to the maintainers, if  you'd like to.

Done! :)

What do you think of the PR?  Is there anything from yours you think we
should incorporate?

File: pkgs/applications/window-managers/wayfire/default.nix
Comment (wucke13) [inline pkgs/applications/window-managers/wayfire/default.nix:49]: Why are `git` and `wayland` in the `nativeBuildInputs`?

File: pkgs/applications/window-managers/wayfire/default.nix
Comment (wucke13) [inline pkgs/applications/window-managers/wayfire/default.nix:55]: Why is this needed?

File: pkgs/applications/window-managers/wayfire/wf-config.nix
Comment (wucke13) [inline pkgs/applications/window-managers/wayfire/wf-config.nix:31]: Out of curiosity, why is `propagatedBuildInputs` used here for the `glm` dep?

File: pkgs/applications/window-managers/wayfire/wf-shell.nix
Comment (wucke13) [inline pkgs/applications/window-managers/wayfire/wf-shell.nix:47]: ~Again, isn't this missing the `wrapGAppsHook`~?
It is not, because of `gnome.gtk` in `buildInputs`?

File: pkgs/top-level/all-packages.nix
Comment (wucke13) [inline pkgs/top-level/all-packages.nix:22521]: I would prefer these to be sorted alphanumerical. Maybe like this:

```
wayfireApplications = wayfireApplications-unwrapped.withPlugins (plugins: [ plugins.wf-shell ]);
inherit (wayfireApplications) wayfire wcm;
wayfireApplications-unwrapped = callPackage ../applications/window-managers/wayfire/applications.nix { };
wayfirePlugins = callPackage ../applications/window-managers/wayfire/plugins.nix {
  inherit (wayfireApplications-unwrapped) wayfire;
};
wf-config = callPackage ../applications/window-managers/wayfire/wf-config.nix { };
```
Also: How does this work without defining `wcm` as a package?

**Review by wucke13 — APPROVED** [review]
I left a few comments and questions. Mostly things that I don't understand :laughing: 
In general, the PR looks good to me. Compilation works, including icons for `wcm` and `wf-shell`

**Comment by wucke13** [comment]
Huh, on running this standalone I lost the icons on wcm. Maybe `wrapGAppsHook` is necessary?

File: pkgs/applications/window-managers/wayfire/wf-config.nix
Comment (alyssais) [reply pkgs/applications/window-managers/wayfire/wf-config.nix:31]: Because any package that has wf-config as an input also needs glm as an input. This would otherwise have to be copied every time something had a wf-config input.

File: pkgs/top-level/all-packages.nix
Comment (alyssais) [reply pkgs/top-level/all-packages.nix:22521]: > Also: How does this work without defining `wcm` as a package?

```nix
{ inherit (wayfireApplications) wayfire wcm; }
```

is equivalent to

```nix
{ wayfire = wayfireApplications.wayfire;
  wcm = wayfireApplications.wcm; }
```

File: pkgs/applications/window-managers/wayfire/default.nix
Comment (alyssais) [reply pkgs/applications/window-managers/wayfire/default.nix:55]: So that Wayfire default configuration files can be placed in /etc. Otherwise people would have no way to set those, because Wayfire would look in its prefix, which can’t be altered.

File: pkgs/applications/window-managers/wayfire/default.nix
Comment (alyssais) [reply pkgs/applications/window-managers/wayfire/default.nix:49]: Wayland was there for wayland-scanner. Not sure why git was there. Seems to work fine without it so I’ll drop it.

File: pkgs/top-level/all-packages.nix
Comment (alyssais) [reply pkgs/top-level/all-packages.nix:22521]: I’ll resort alphabetically :)

**Comment by alyssais** [comment]
Sorry for being slow with this.

> Huh, on running this standalone I lost the icons on wcm. Maybe `wrapGAppsHook` is necessary?

I tried adding that to both wf-shell and wcm, but still no icons. Any other ideas? Did it work in your PR?

> Also: We should look at adding some facility for launching `dbus` to the package. Otherwise things like the gnome pinentry won't work.

Good idea. I’ll have a look at the sway wrapper and see what goodies we can steal.

**Comment by wucke13** [comment]
> Sorry for being slow with this.

No worries!

> I tried adding that to both wf-shell and wcm, but still no icons. Any other ideas? Did it work in your PR?

In my PR all icons but the one on the "Back" icon work.

> Good idea. I’ll have a look at the sway wrapper and see what goodies we can steal.

Sweet!

**Comment by wucke13** [comment]
@alyssais 
What about the [new release](https://github.com/WayfireWM/wayfire/releases/tag/0.5.0)? Can you update your PR accordingly?

**Comment by alyssais** [comment]
>> I tried adding that to both wf-shell and wcm, but still no icons. Any other ideas? Did it work in your PR?
>
> In my PR all icons but the one on the "Back" icon work.

I tried out your PR, and while icons for individual settings views in
WCM, the header icons didn't.

I managed to get this PR to the same functionality, but it's still not
quite right.

**Comment by wucke13** [comment]
@alyssais What icons are missing? I think the issue which I meant is solved in your PR:

**Comment by alyssais** [comment]
> @alyssais What icons are missing? I think the issue which I meant is
> solved in your PR:

Interesting!  I don't get the "General", "Accessibility", etc. icons.
Do you have GNOME enabled?

**Comment by wucke13** [comment]
No, but I had it installed. So maybe it did some GTK RC magic which fixes the issue?

**Comment by alyssais** [comment]
I've realised that I have the same problem with other GTK applications,
so it's not a Wayfire issue and therefore doesn't need to hold up this
PR.

As for the D-Bus stuff, it's really not an area I know anything about,
so I think it would be best addressed by somebody else as a follow-up to
this PR.

So my plan here I think is to update to the newest Wayfire release, and
then if there's nothing else outstanding I'll merge not long afterward.

File: pkgs/applications/window-managers/wayfire/default.nix
Comment (diamondburned) [inline pkgs/applications/window-managers/wayfire/default.nix:12]: I seem to get a hash mismatch with this:

```
unpacking source archive /build/v0.6.0.tar.gz
hash mismatch in fixed-output derivation '/nix/store/lbyaykpqg15k20v18qvmmzyar7vdd3w6-source':
  wanted: sha256:1b22gymqfn0c49nf39676q5bj25rxab874iayiq31cmn14r30dyg
  got:    sha256:0xdjrmpdwag2d7bhqa2iall0rdsqpzfgix279lp2zy1ldw9gzg1n
```

File: pkgs/applications/window-managers/wayfire/wcm.nix
Comment (diamondburned) [inline pkgs/applications/window-managers/wayfire/wcm.nix:13]: I seem to get a hash mismatch with this:

```
unpacking source archive /build/v0.6.0.tar.gz
hash mismatch in fixed-output derivation '/nix/store/lbyaykpqg15k20v18qvmmzyar7vdd3w6-source':
  wanted: sha256:1b22gymqfn0c49nf39676q5bj25rxab874iayiq31cmn14r30dyg
  got:    sha256:0xdjrmpdwag2d7bhqa2iall0rdsqpzfgix279lp2zy1ldw9gzg1n
```

**Pull request merged**
